### ISSUE TITLE
Improve feedback for stored procedure errors

### ISSUE THREAD

**Issue by thomaslevesque** [Azure/azure-documentdb-dotnet#518]
When a call to a stored procedure fails because of an error in the SP, the response is always 400 (BadRequest), which doesn't convey any useful information as to why the request failed. The response body looks like this:

```json
{"Errors":["Encountered exception while executing function. Exception = Error: {\"Errors\":[\"Resource Not Found\"]}\r\nStack trace: Error: {\"Errors\":[\"Resource Not Found\"]}\n   at Anonymous function (myStoredProc.js:13:22)\n   at Anonymous function (myStoredProc.js:515:29)"]}
```

The way the error is formatted makes it unnecessarily hard to parse (mixed JSON and plain text inside a JSON string), and even if we managed to parse it, the only useful thing we can extract from it is "Resource Not Found", which is usable, but an error code would be better.

In addition, the response contains an undocumented header, `x-ms-substatus`, which contains a more useful value (404 in my case). But there's no easy way to get it from the exception, short of parsing it from the `DocumentClient.ResponseHeaders` property. Apparently `DocumentClient` has a `GetSubStatus` method, but it's internal. Making it public would be very helpful.

**Comment by ealsur** [comment]
Would this helper solve the issue? https://github.com/Azure/azure-documentdb-changefeedprocessor-dotnet/blob/7afeeb16ed16e7730e9622a36b063951a8c55fe2/src/DocumentDB.ChangeFeedProcessor/DocDBErrors/SubStatusHelpers.cs

**Comment by ealsur** [comment]
I'm not sure how to return a specific HTTP status code from within a sproc, @mkolt, is there a way to do it?

**Comment by mkolt** [comment]
Returning settable status code as part of script execution is possible, but would be wrong. Even if script fails, changing failure status code would be at least confusing and hard for diagnostics. While 400 for the request with settable substatus is deterministic, on the client side getting the substatus is not a big deal at all. Does this make sense?

**Comment by thomaslevesque** [comment]
> Does this make sense?

I guess. But it's a pity there's no easy way to return more details about the error.

**Comment by veleek** [comment]
@mkolt - There are valid reasons for wanting to set the status code.  By default, if an error is unhandled in some way, then overriding the status code to 400 probably might be reasonable (but would be nice if it was configurable).  But in scenarios where the error is explicit (e.g. I actually call `__.response.setStatusCode` or something), I don't think there's any reason we should want to prevent this.  The sproc successfully completed, and the result was that there was a conflict.

**Comment by mkolt** [comment]
If we allowed from inside sproc to set request status code, it would be quite a mess as at request level status codes are used in metrics, to trigger internal alerts, etc.  Status for all requests including sproc need to be consistent and manageable. In particular, an error in sproc means invalid payload (400), more details can be obtained from substatus and error message. So, if an error code is needed on the client,  just use substatus code, If more details are needed, use error message. I.e. use throw new Error(substatusCode, "error message").

**Comment by veleek** [comment]
What about for non-error cases?  If I'm not throwing an error, is there any way for me to differentiate between a 202 Created, and a 200 OK, or maybe a 304 not modified without needing to return an arbitrary object back from my request?

**Comment by mkolt** [comment]
Well, I think it's similar issue: script is a wrapper, so status is for the wrapper itself. Using custom return object shouldn't be a big deal, right?

**Comment by veleek** [comment]
It's not a huge deal.  It's just a little bit of a disconnect in how sprocs
are written or used.  From the calling side, when I call Update I have to
handle a response differently than when I call my UpdateUsingSproc method.
Inside my sprocs, I need to handle methods that may return a different
status differently than methods that always return the same thing.  And
since there's no easy way to share code between sprocs, if I want every
sproc to return the same "object" I need to copy helpers for creating those
response objects across each sproc even though in 90% of them I'm going to
be using a constant 200 value for status.

I totally understand that there are valid reasons for wanting to be able to
depend on the status code for internal reporting and metrics.  But at the
same time since all errors from a sproc are handled the same way it seems
like it would be pretty easy to differentiate between various statuses for
reporting purposes:
* Did the sproc return an error? Record it as a 400 internally for
reporting and metrics.  Then return the response using the status code
provided by the user.
* Did the sproc succeed?  Record a 200 and return the status code provided
by the user.

There's no way for my sproc to return anything other than a 200 or a 400
(in regular circumstances), so what sort of information am I hiding by
changing the status code?  Any failure that wouldn't result in a 400 (e.g.
that didn't come from my sproc "throw"ing something), I would expect a 500
to get returned anyway so that'll still get reported.
